derp/derp_client.go

```diff
@@ -524,7 +532,16 @@ func (c *Client) recvTimeout(timeout time.Duration) (m ReceivedMessage, err erro
 				c.logf("[unexpected] dropping short peerGone frame from DERP server")
 				continue
 			}
-			pg := PeerGoneMessage(key.NodePublicFromRaw32(mem.B(b[:keyLen])))
+			// Backward compatibility for the older peerGone version without reason byte
+			var reason byte = peerGoneReasonDisconnected
+			if n != keyLen {
+				reason = b[keyLen]
+			}
+			if reason == peerGoneReasonUnknown {
+				pg := PeerGoneMessageUnknown(key.NodePublicFromRaw32(mem.B(b[:keyLen])))
+				return pg, nil
+			}
+			pg := PeerGoneMessageDisconnected(key.NodePublicFromRaw32(mem.B(b[:keyLen])))
```

Comment:
If we add more reason codes later they will end up classified here until the clients update.

Reply:
Thank you, this is one of the places where I am not sure what is the most forward/backward compatible way. Fundamentally, it comes down to this question:

Should future new peerGone reasons cause the client to log an informative error message or not?

The current code preserves the old behavior (no log message) for new reason codes. The alternative would cause the new behavior (log a message) for all future new codes.

I am not convinced that either is correct so happy to hear other people's thoughts!

Reply:
I ended up pushing the gone reason out to the client so it could decide what to do. Currently the clients log all the reasons other than disconnected.